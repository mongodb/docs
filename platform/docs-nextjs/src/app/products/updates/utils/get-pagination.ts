/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
