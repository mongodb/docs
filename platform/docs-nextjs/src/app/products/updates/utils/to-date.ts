export const normalizeDate = (date: string) => {
  // Check if the date string has the correct format 'YYYY-MM-DD'
  const dateToNormalize = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date)?.slice(1); // excludes the full date in the first index i.e. "2025-05-27"
  if (dateToNormalize) {
    // This is needed because, JavaScript’s Date constructor uses zero-based months (January = 0 … December = 11).
    return new Date(Number(dateToNormalize[0]), Number(dateToNormalize[1]) - 1, Number(dateToNormalize[2]));
  }
  return new Date(date);
};
