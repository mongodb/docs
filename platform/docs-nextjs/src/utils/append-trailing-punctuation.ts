import { ASTNode, ReferenceNode, RefRoleNode } from "@/types/ast";
import { isReferenceNode, isRefRoleNode, isTextNode } from "@/types/ast-utils";

// Append dangling punctuation to preceding Link's text
export const appendTrailingPunctuation = (nodes: ASTNode[]) => {
  const truncatedNodes = [];

  for (let i = 0; i < nodes.length; i++) {
    const currNode = nodes[i];
    const nextNode = nodes[i + 1];
    const hasDanglingSibling =
      isTextNode(nextNode) &&
      nextNode.value.length === 1 &&
      !nextNode.value.match(/\s/);
    // TODO: if issues arise, instead of matching for non-whitespace, we might consider checking FOR certain singular punctuation
    // ie: regex match for  .?!,:;)*"']}`

    if ((isReferenceNode(currNode) || isRefRoleNode(currNode)) && hasDanglingSibling) {
      const copyOfNodeWithDeepCopiedChildren = deepCopyNodesChildren(currNode);
      copyOfNodeWithDeepCopiedChildren.children.push(nextNode);
      truncatedNodes.push(copyOfNodeWithDeepCopiedChildren);
      i++;
      continue;
    }
    truncatedNodes.push(currNode);
  }
  return truncatedNodes;
};

// Make a copy of node with a deep copy of the new child node
function deepCopyNodesChildren(refNode: ReferenceNode | RefRoleNode) {
  if (!refNode.children || !refNode.children.length) return { ...refNode };
  const copyOfChildren = [...refNode.children];

  const copyOfNode = {
    ...refNode,
    children: copyOfChildren,
  };
  return copyOfNode;
}
