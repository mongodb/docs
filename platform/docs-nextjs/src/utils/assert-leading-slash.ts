const assertLeadingSlash = (url: string) => {
  if (url && url.match(/^\//)) {
    return url;
  }
  return `/${url}`;
};

export { assertLeadingSlash };
