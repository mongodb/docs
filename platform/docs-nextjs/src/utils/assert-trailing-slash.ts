export const assertTrailingSlash = (url: string) => {
  if (url && url.match(/\/$/)) {
    return url;
  }
  return `${url}/`;
};
