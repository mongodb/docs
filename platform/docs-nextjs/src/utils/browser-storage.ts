import { isBrowser } from './is-browser';

const isValidStorage = isBrowser;

export const setLocalValue = (key: string, value: unknown) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      localStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting localStorage value');
  }
};

export const getLocalValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting localStorage value');
    return undefined;
  }
};

export const setSessionValue = (key: string, value: string) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      sessionStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting session value');
  }
};

export const getSessionValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting session value');
    return undefined;
  }
};
