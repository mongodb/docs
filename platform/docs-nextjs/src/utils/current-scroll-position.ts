import { isBrowser } from './is-browser';

// Function that returns current scroll position for analytics
export const currentScrollPosition = (): number => {
  if (isBrowser) {
    const percentage = Math.round((window.scrollY / (document.body.scrollHeight - window.innerHeight)) * 100);
    return Math.round(percentage / 10) * 10;
  }
  return 0;
};
