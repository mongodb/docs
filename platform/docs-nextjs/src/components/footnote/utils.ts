import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
