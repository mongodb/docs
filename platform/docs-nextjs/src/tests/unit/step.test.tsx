import { render } from '@testing-library/react';
import { mockLocation } from '../utils/mock-location';
import Step from '@/components/procedure/step';
// data for this component
import mockData from '../data/step.test.json';
import type { StepNode } from '@/types/ast';

beforeAll(() => {
  mockLocation({ hash: '' });
});

it.skip('renders with "connected" styling by default', () => {
  const tree = render(<Step nodeChildren={(mockData as StepNode).children} stepNumber={1} />);
  expect(tree.asFragment()).toMatchSnapshot();
});

it.skip('renders with "connected" styling', () => {
  const tree = render(<Step nodeChildren={(mockData as StepNode).children} stepStyle="connected" stepNumber={1} />);
  expect(tree.asFragment()).toMatchSnapshot();
});

it.skip('renders with "normal" or YAML steps styling', () => {
  const tree = render(<Step nodeChildren={(mockData as StepNode).children} stepStyle="normal" stepNumber={1} />);
  expect(tree.asFragment()).toMatchSnapshot();
});
