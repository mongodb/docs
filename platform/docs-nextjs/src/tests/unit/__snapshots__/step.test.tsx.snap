// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with "connected" styling 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  gap: 33px;
}

.emotion-0 h2,
.emotion-0 h3,
.emotion-0 h4 {
  margin-top: 4px;
}

.emotion-0:not(:last-child):before {
  content: '';
  border-left: 2px dashed #E8EDEB;
  bottom: 0;
  left: 16px;
  position: absolute;
  top: 0;
}

.dark-theme .emotion-0:not(:last-child):before {
  border-left-color: #5C6C75;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  position: relative;
  min-height: 59px;
}

.emotion-3 {
  position: relative;
  font-weight: bold;
  background-color: #E3FCF7;
  color: #00684A;
  height: 34px;
  width: 34px;
}

.dark-theme .emotion-3 {
  background-color: #00684A;
  color: #E8EDEB;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 {
  padding-bottom: 64px;
}

@media only screen and (max-width: 767px) {
  .emotion-5 {
    padding-bottom: 40px;
  }
}

@media only screen and (max-width: 480px) {
  .emotion-5 {
    padding-bottom: 32px;
  }
}

.emotion-6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 0;
}

.emotion-7 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-7 strong,
.emotion-7 b {
  font-weight: 700;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-decoration-color: transparent;
  line-height: 13px;
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-9>span>code,
.emotion-9>code {
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-9:focus,
.emotion-9:hover {
  text-decoration-line: underline;
  -webkit-transition: text-decoration 150ms ease-in-out;
  transition: text-decoration 150ms ease-in-out;
  text-underline-offset: 4px;
  text-decoration-thickness: 2px;
}

.emotion-9:focus {
  text-decoration-color: #016BF8;
  outline: none;
}

.emotion-9:hover {
  text-decoration-color: #E8EDEB;
}

<div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3 emotion-4"
      >
        1
      </div>
    </div>
    <div
      class="emotion-5 emotion-6"
    >
      <p
        class="emotion-7"
      >
        Import data from CSV or JSON files into your MongoDB database.
      </p>
      <p
        class="emotion-7"
      >
        <a
          class="emotion-9"
          href="/import-export/#std-label-compass-import-export"
        >
          To learn more, see Import and Export Data
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with "connected" styling by default 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  gap: 33px;
}

.emotion-0 h2,
.emotion-0 h3,
.emotion-0 h4 {
  margin-top: 4px;
}

.emotion-0:not(:last-child):before {
  content: '';
  border-left: 2px dashed #E8EDEB;
  bottom: 0;
  left: 16px;
  position: absolute;
  top: 0;
}

.dark-theme .emotion-0:not(:last-child):before {
  border-left-color: #5C6C75;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  position: relative;
  min-height: 59px;
}

.emotion-3 {
  position: relative;
  font-weight: bold;
  background-color: #E3FCF7;
  color: #00684A;
  height: 34px;
  width: 34px;
}

.dark-theme .emotion-3 {
  background-color: #00684A;
  color: #E8EDEB;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 {
  padding-bottom: 64px;
}

@media only screen and (max-width: 767px) {
  .emotion-5 {
    padding-bottom: 40px;
  }
}

@media only screen and (max-width: 480px) {
  .emotion-5 {
    padding-bottom: 32px;
  }
}

.emotion-6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 0;
}

.emotion-7 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-7 strong,
.emotion-7 b {
  font-weight: 700;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-decoration-color: transparent;
  line-height: 13px;
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-9>span>code,
.emotion-9>code {
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-9:focus,
.emotion-9:hover {
  text-decoration-line: underline;
  -webkit-transition: text-decoration 150ms ease-in-out;
  transition: text-decoration 150ms ease-in-out;
  text-underline-offset: 4px;
  text-decoration-thickness: 2px;
}

.emotion-9:focus {
  text-decoration-color: #016BF8;
  outline: none;
}

.emotion-9:hover {
  text-decoration-color: #E8EDEB;
}

<div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3 emotion-4"
      >
        1
      </div>
    </div>
    <div
      class="emotion-5 emotion-6"
    >
      <p
        class="emotion-7"
      >
        Import data from CSV or JSON files into your MongoDB database.
      </p>
      <p
        class="emotion-7"
      >
        <a
          class="emotion-9"
          href="/import-export/#std-label-compass-import-export"
        >
          To learn more, see Import and Export Data
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with "normal" or YAML steps styling 1`] = `
<DocumentFragment>
  .emotion-0 {
  gap: 16px;
}

.emotion-0 h2,
.emotion-0 h4,
.emotion-0 h5,
.emotion-0 h6 {
  margin-top: unset;
}

.emotion-0 h3 {
  margin-top: 2px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  position: relative;
  min-height: 59px;
}

.emotion-3 {
  border: 1.5px solid #C1C7C6;
  font-weight: 500;
  font-size: 16px;
  line-height: 28px;
  background-color: inherit;
  color: var(--opposite-color);
  height: 27px;
  width: 27px;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 50%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-5 section>*,
.emotion-5 ol p,
.emotion-5 ul p {
  margin-bottom: 16px;
}

.emotion-6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 0;
}

.emotion-7 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-7 strong,
.emotion-7 b {
  font-weight: 700;
}

.emotion-9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-decoration-color: transparent;
  line-height: 13px;
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-9>span>code,
.emotion-9>code {
  color: var(--link-color-primary);
  font-weight: var(--link-font-weight);
}

.emotion-9:focus,
.emotion-9:hover {
  text-decoration-line: underline;
  -webkit-transition: text-decoration 150ms ease-in-out;
  transition: text-decoration 150ms ease-in-out;
  text-underline-offset: 4px;
  text-decoration-thickness: 2px;
}

.emotion-9:focus {
  text-decoration-color: #016BF8;
  outline: none;
}

.emotion-9:hover {
  text-decoration-color: #E8EDEB;
}

<div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3 emotion-4"
      >
        1
      </div>
    </div>
    <div
      class="emotion-5 emotion-6"
    >
      <p
        class="emotion-7"
      >
        Import data from CSV or JSON files into your MongoDB database.
      </p>
      <p
        class="emotion-7"
      >
        <a
          class="emotion-9"
          href="/import-export/#std-label-compass-import-export"
        >
          To learn more, see Import and Export Data
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
